sklearn/utils/validation.py

```diff
@@ -966,3 +967,134 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+
+    - that there are no significant negative eigenvalues (with absolute value
+      more than 1e-10 and more than 1e-5 times the largest positive
+      eigenvalue). If this check fails, it raises a `PSDSpectrumWarning`. All
+      negative eigenvalues (even smaller ones) are set to zero in all cases.
+
+    - that the eigenvalues are well conditioned. That means, that the
+      eigenvalues are all greater than the maximum eigenvalue divided by 1e12.
+      If this check fails and `warn_on_zeros=True`, it raises a
+      `PSDSpectrumWarning`. All the eigenvalues that are too small are then set
+      to zero.
+
+    Parameters
+    ----------
+    lambdas : array-like
+        Array of eigenvalues to check / fix.
+
+    warn_on_zeros : boolean (default: False)
+        When this is set to `True`, a `PSDSpectrumWarning` will be raised when
+        there are extremely small eigenvalues. Otherwise no warning will be
+        raised. Note that in both cases, extremely small eigenvalues will be
+        set to zero.
+
+    Returns
+    -------
+    lambdas_fixed : array-like
+        A fixed validated copy of the array of eigenvalues.
+
+    Examples
+    --------
+    >>> check_psd_eigenvalues((1, 2))      # nominal case
+    ... # doctest: +NORMALIZE_WHITESPACE
+    array([1, 2])
+    >>> check_psd_eigenvalues((5, 5j))     # significant imag part
+    ... # doctest: +ELLIPSIS +NORMALIZE_WHITESPACE
+    Traceback (most recent call last):
+        ...
+    ValueError: There are significant imaginary parts in eigenvalues (1.000000
+        of the max real part). The matrix is maybe not PSD, or something went
+        wrong with the eigenvalues decomposition.
+    >>> check_psd_eigenvalues((5, 5e-5j))  # insignificant imag part
+    ... # doctest: +NORMALIZE_WHITESPACE
+    array([5., 0.])
+    >>> check_psd_eigenvalues((-5, -1))    # all negative
+    ... # doctest: +ELLIPSIS +NORMALIZE_WHITESPACE
+    Traceback (most recent call last):
+        ...
+    ValueError: All eigenvalues are negative (max is -1.000000). The matrix is
+        maybe not PSD, or something went wrong with the eigenvalues
+        decomposition.
+    >>> check_psd_eigenvalues((5, -1))     # significant negative
+    ... # doctest: +NORMALIZE_WHITESPACE
+    array([5, 0])
+    >>> check_psd_eigenvalues((5, -5e-5))  # insignificant negative
+    ... # doctest: +NORMALIZE_WHITESPACE
+    array([5., 0.])
+    >>> check_psd_eigenvalues((5, 4e-12))  # bad conditioning (too small)
+    ... # doctest: +NORMALIZE_WHITESPACE
+    array([5., 0.])
+
+    """
+
+    # Check that there are no significant imaginary parts
+    if not np.isreal(lambdas).all():
+        max_imag_abs = abs(np.imag(lambdas)).max()
+        max_real_abs = abs(np.real(lambdas)).max()
+        if max_imag_abs > 1e-5 * max_real_abs:
+            raise ValueError(
+                "There are significant imaginary parts in eigenvalues (%f "
+                "of the max real part). The matrix is maybe not PSD, or "
```

Comment:
for max -> ok.

for the percentage symbol:
well, since it is 1e-5  that's 0.00005 so 0.005% . Not sure that the percentage style helps here, but if you think that improves readability I can certainly do it.